### PR TITLE
qemu-guest-agent: Add new case for operate ssh-pub-keys into guest.

### PR DESCRIPTION
**Pull request by 6-dehan**
operation includes guest-ssh-add-authorized-keys, 
*-remove-*, *-get-*, and reset ssh keys" for host can 
login guest via ssh without passwd.

ID: 1906644
Signed-off-by: Dehan Meng <demeng@redhat.com>

### CODE REVIEW

**Comment by 6-dehan** [comment]
@vivianQizhu please help to review this, thanks.
(1/1) type_specific.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection: PASS (60.16 s)

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -1176,6 +1176,124 @@ def get_allocation_bitmap():
...
+        _login_guest_test(self, session, guest_ip_ipv4)
+
+        error_context.context("Check whether can reset keys", logging.info)
+        host_key4 = _generate_host_keys(self)
+        self.gagent.ssh_add_authorized_keys("fedora", host_key4, reset=True)
+        keys_after_reset_ga = self.gagent.ssh_get_authorized_keys("fedora")
+        session.cmd(params["add_line_at_end"] % guest_path)
+        keys_after_reset_guest = session.cmd_output(cmd_get_guest_keys).strip()
+        _value_compared_ga_guest(self, keys_after_reset_ga,
+                                 keys_after_reset_guest, "reset")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1291]: Seems the above part could be abstract into a function like:
```
def ssh_key_test(operation, *keys, reset=False):
    self.gagent.ssh_add_authorized_keys("fedora", *keys, reset=reset)
    keys_ga = self.gagent.ssh_get_authorized_keys("fedora")
    session.cmd(params["add_line_at_end"] % guest_path)
    keys_guest = session.cmd_output(cmd_get_guest_keys).strip()
    _value_compared_ga_guest(self, keys_after_reset_ga,
￼                                 keys_after_reset_guest, operation)
    return keys_ga, keys_guest
```

**Comment by 6-dehan** [comment]
@kimi1978wy 
 (1/1) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection: PASS (63.39 s)

**Comment by 6-dehan** [comment]
> @kimi1978wy
> (1/1) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection: PASS (63.39 s)

@vivianQizhu

**Comment by vivianQizhu** [comment]
Seems your code is dirty now, `check_get_disks` should not appear, please check it.

**Comment by 6-dehan** [comment]
> Seems your code is dirty now, `check_get_disks` should not appear, please check it.

conflict issues, not check it clearly, my bad, updated it just right now.

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 2]
```diff
@@ -286,6 +286,18 @@
...
+            gagent_check_type = get_disks
+            force_create_image_stg0 = yes
+            remove_image_stg0 = yes
+            images += " stg0"
+            image_name_stg0 = images/storage0
+            image_size_stg0 = 1G
+            get_rawdisk_guest_cmd = lsblk -p|grep disk|awk '{print$1}'
+            diskinfo_guest_cmd = 'lsblk -Jp -o KNAME,PKNAME,TYPE | grep %s|awk "NR==1"'
+            rawdisk_info_guest_cmd = 'lsblk -Jp -o KNAME,TYPE | grep %s|awk "NR==1"'
+            cmd_get_disk_alias = 'lsblk -o NAME %s |grep -v NAME'
```
Comment (vivianQizhu) [inline qemu/tests/cfg/qemu_guest_agent.cfg:300]: The commit is still not correct?

**Comment by 6-dehan** [comment]
@vivianQizhu, issue has gone.

**Comment by 6-dehan** [comment]
@vivianQizhu 
(1/1) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection.q35: PASS (63.13 s)

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -1291,6 +1291,115 @@ def gagent_check_get_disks(self, test, params, env):
...
+            status = process.system(params["ssh_keygen_cmd"],
+                                    shell=True, timeout=3)
+            if status:
+                test.error("Can not generate ssh key with no "
+                           "interaction, please have a check.")
+            cmd_get_hostkey = params["cmd_get_hostkey"]
+            host_key = process.system_output(cmd_get_hostkey,
+                                             shell=True, timeout=3)
+            host_key = host_key.decode(encoding="utf-8",
+                                       errors="strict")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1339]:
```suggestion
            host_key = process.getoutput(cmd_get_hostkey,
                                         shell=True, timeout=3)
```

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -1291,6 +1291,115 @@ def gagent_check_get_disks(self, test, params, env):
...
+            return host_key
+
+        def _login_guest_test(guest_ip):
+            cmd_login_guest = params["test_login_guest"] % guest_ip
+            login_user = process.system_output(cmd_login_guest,
+                                               shell=True, timeout=5)
+            login_user = login_user.strip().decode(encoding="utf-8",
+                                                   errors="strict")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1347]:
```suggestion
            login_user = process.getoutput(cmd_login_guest,
                                           shell=True, timeout=5).strip()
```

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -1291,6 +1291,115 @@ def gagent_check_get_disks(self, test, params, env):
...
+        os_type = self.params["os_type"]
+        guest_ip_ipv4 = utils_net.get_guest_ip_addr(session, mac_addr,
+                                                    os_type)
+
+        error_context.context("Check the basic function ",
+                              logging.info)
+        host_key1 = _generate_host_keys()
+        cmd_guest_keys = params["cmd_cat_key_file"] % guest_home_path
+        ssh_key_test("add", host_key1, reset=False)
+        _login_guest_test(guest_ip_ipv4)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1375]: I suppose you don't need this function, would you like to try this:
```
params['password'] == ''
vm.wait_for_login(username=params["guest_user"], password='')
```

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -1291,6 +1291,115 @@ def gagent_check_get_disks(self, test, params, env):
...
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+
+        def ssh_key_test(operation, *keys, **kwargs):
+            guest_user = params['guest_user']
+            op_func = getattr(self.gagent, 'ssh_%s_authorized_keys' % operation)
+            op_func(guest_user, *keys, **kwargs)
+            keys_ga = self.gagent.ssh_get_authorized_keys("fedora")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1315]:
```suggestion
            keys_ga = self.gagent.ssh_get_authorized_keys(guest_user)
```

**Review by vivianQizhu — COMMENTED** [review]
And please add docstring for all the functions.

**Comment by 6-dehan** [comment]
@vivianQizhu, 
(1/1) Host_RHEL.m8.u4.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection.q35: PASS (67.66 s)

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -1291,6 +1291,102 @@ def gagent_check_get_disks(self, test, params, env):
...
+        self._open_session_list.append(session)
+        _prepared_n_restore_env("0")
+        cmd_guest_keys = params["cmd_get_guestkey"]
+
+        error_context.context("Check the basic function ",
+                              logging.info)
+        host_key1 = _generate_host_keys()
+        ssh_key_test("add", host_key1, reset=False)
+        self.vm.wait_for_login(username=params["guest_user"],
+                               password=params["guest_user_passwd"])
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1360]: I guess this is not what you want, you should want to login it without the password, so should make password = "", isn't it?

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -1291,6 +1291,102 @@ def gagent_check_get_disks(self, test, params, env):
...
+            keys_ga = self.gagent.ssh_get_authorized_keys(guest_user)
+            session.cmd(params["add_line_at_end"])
+            keys_guest = session.cmd_output(cmd_guest_keys).strip()
+            _value_compared_ga_guest(keys_ga, keys_guest, operation)
+            return keys_ga, keys_guest
+
+        def _prepared_n_restore_env(status):
+            if status != "1":
+                session.cmd(params["cmd_add_user_set_passwd"])
+                session.cmd(params["set_setenforce"] % int(status))
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1324]: Seems `status` now is meaningless here, recommend:
```suggestion
        def _prepared_n_restore_env(prepare=True):
            if prepare:
                session.cmd(params["cmd_add_user_set_passwd"])
                session.cmd(params["set_setenforce"] % 0)
```

**Comment by vivianQizhu** [comment]
And again, docstring please. Thanks.

**Review by 6-dehan — COMMENTED** [review]
> And again, docstring please. Thanks.

oh, sorry, I missed that. adding now.

File: qemu/tests/qemu_guest_agent.py (hunk 7 above)
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:1360]: Are there any special rules for this method use? I've tried several times with this function,but still gave me this error: "Got password prompt twice    (output: "\nPermission denied, please try again.\nfedora@10.73.73.168's password: ")". host should login guest without password definitely. but it just goes against the wish. still give it more try. just update resent result here. if you can give me some advice that's more than enough for me.thanks

Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py:1360]: @6-dehan Sorry, I just checked the method `remote.remote_login()`, which is called by `vm.wait_for_login()`, there it adds the option `-o PreferredAuthentications=password` when send `ssh` command, so even you generate keys it still asks for the password. Please go back to your original method, sorry for the inconvenience.

**Comment by 6-dehan** [comment]
@vivianQizhu update it now, focus on other tasks recently, please have a check when you're free. thanks
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_ssh_public_key_injection.q35: PASS (64.68 s)

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py:1409]: @6-dehan Please check the above comment, thanks.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:1409]: Idk exactly why it always get fails when I used 'process.getoutput', but 'process.system_output' can be passed.

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -1354,6 +1354,147 @@ def gagent_check_get_disks(self, test, params, env):
...
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+
+        def ssh_key_test(operation, *keys, **kwargs):
+            """
+            Do ssh_key generation and ssh_login test to confirm
+            whether the ssh_key api work well or not.
+
+            :param operation:
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:1380]: Should not be empty.

**Review by vivianQizhu — COMMENTED** [review]
LGTM otherwise.

File: qemu/tests/qemu_guest_agent.py (hunk 9 above)
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py]: sorry, my bad, new code has been updated.

**Pull request merged**